In [92]:

# importamos las librerías que necesitamos

# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# Visualización
# ------------------------------------------------------------------------------
import matplotlib.pyplot as plt
import seaborn as sns

# Evaluar linealidad de las relaciones entre las variables
# y la distribución de las variables
# ------------------------------------------------------------------------------
import scipy.stats as stats
from scipy.stats import chi2_contingency, ttest_ind

# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

# Gestión de los warnings
# -----------------------------------------------------------------------
import warnings
warnings.filterwarnings("ignore")

# PDTE: IMPORTAR SOPORTE CUANDO CORRESPONDA 🚩

In [93]:
# cargamos el dataframe correspondiente a Customer Flight Analysis.csv
df_vuelos = pd.read_csv("../archivos/Customer-Flight-Activity.csv") # De momento mantengo Loyalty Number como columna 🚨
df_vuelos = df_vuelos.sort_values(by=['Loyalty Number', 'Year', 'Month'], ascending=[True, True, True]).reset_index(drop=True)
display(df_vuelos.head())
display(df_vuelos.tail())
display(df_vuelos.sample(5))

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
0          100018  2017      1               3                        0   
1          100018  2017      2               2                        2   
2          100018  2017      3              14                        3   
3          100018  2017      4               4                        0   
4          100018  2017      5               0                        0   

   Total Flights  Distance  Points Accumulated  Points Redeemed  \
0              3      1521               152.0                0   
1              4      1320               132.0                0   
2             17      2533               253.0              438   
3              4       924                92.0                0   
4              0         0                 0.0                0   

   Dollar Cost Points Redeemed  
0                            0  
1                            0  
2                           36  
3                            0  
4                            0

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
405619          999986  2018      8               8                        1   
405620          999986  2018      9               0                        0   
405621          999986  2018     10               6                        0   
405622          999986  2018     11              11                        4   
405623          999986  2018     12               0                        0   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
405619              9      3672               367.0              655   
405620              0         0                 0.0                0   
405621              6       708                70.0                0   
405622             15      2040               204.0                0   
405623              0         0                 0.0                0   

        Dollar Cost Points Redeemed  
405619                           53  
405620                            0  
405621                            0  
405622                            0  
405623                            0

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
170044          478278  2017      5               0                        0   
236169          622610  2017     10               0                        0   
99422           322501  2018      3               0                        0   
95027           312774  2017     12               3                        3   
325958          822653  2018      3               0                        0   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
170044              0         0                 0.0                0   
236169              0         0                 0.0                0   
99422               0         0                 0.0                0   
95027               6      2580               258.0              631   
325958              0         0                 0.0                0   

        Dollar Cost Points Redeemed  
170044                            0  
236169                            0  
99422                             0  
95027                            51  
325958                            0

In [94]:
# cargamos el dataframe correspondiente a Customer-Loyalty-History.csv
df_clientes = pd.read_csv("../archivos/Customer-Loyalty-History.csv") # De momento mantengo Loyalty Number como columna 🚨
df_clientes = df_clientes.sort_values(by='Loyalty Number', ascending=True).reset_index(drop=True)
display(df_clientes.head())
display(df_clientes.tail())
display(df_clientes.sample(5))

Loyalty Number Country          Province          City Postal Code  Gender  \
0          100018  Canada           Alberta      Edmonton     T9G 1W3  Female   
1          100102  Canada           Ontario       Toronto     M1R 4K3    Male   
2          100140  Canada  British Columbia  Dawson Creek     U5I 4F1  Female   
3          100214  Canada  British Columbia     Vancouver     V5R 1W3    Male   
4          100272  Canada           Ontario       Toronto     P1L 8X8  Female   

  Education   Salary Marital Status Loyalty Card      CLV Enrollment Type  \
0  Bachelor  92552.0        Married       Aurora  7919.20        Standard   
1   College      NaN         Single         Nova  2887.74        Standard   
2   College      NaN       Divorced         Nova  2838.07        Standard   
3  Bachelor  63253.0        Married         Star  4170.57        Standard   
4  Bachelor  91163.0       Divorced         Star  6622.05        Standard   

   Enrollment Year  Enrollment Month  Cancellation Year  Cancellation Month  
0             2016                 8                NaN                 NaN  
1             2013                 3                NaN                 NaN  
2             2016                 7                NaN                 NaN  
3             2015                 8                NaN                 NaN  
4             2014                 1                NaN                 NaN

Loyalty Number Country          Province         City Postal Code  \
16732          999902  Canada           Ontario      Toronto     M1R 4K3   
16733          999911  Canada      Newfoundland   St. John's     A1C 6H9   
16734          999940  Canada            Quebec  Quebec City     G1B 3L5   
16735          999982  Canada  British Columbia     Victoria     V10 6T5   
16736          999986  Canada           Ontario       Ottawa     K1F 2R2   

       Gender Education    Salary Marital Status Loyalty Card       CLV  \
16732    Male   College       NaN        Married       Aurora   7290.07   
16733    Male    Doctor  217943.0         Single         Nova   8564.77   
16734  Female  Bachelor   47670.0        Married         Nova  20266.50   
16735    Male   College       NaN        Married         Star   2631.56   
16736  Female  Bachelor   46594.0        Married         Nova   8257.01   

      Enrollment Type  Enrollment Year  Enrollment Month  Cancellation Year  \
16732        Standard             2014                 5                NaN   
16733        Standard             2012                 8                NaN   
16734        Standard             2017                 7                NaN   
16735        Standard             2018                 7                NaN   
16736  2018 Promotion             2018                 2                NaN   

       Cancellation Month  
16732                 NaN  
16733                 NaN  
16734                 NaN  
16735                 NaN  
16736                 NaN

Loyalty Number Country          Province            City Postal Code  \
2040           209675  Canada            Quebec        Montreal     H4G 3T4   
6123           430500  Canada          Manitoba        Winnipeg     R2C 0M5   
10844          682419  Canada            Quebec       Tremblant     H5Y 2S9   
5291           385190  Canada            Quebec        Montreal     H2Y 4R4   
11080          694915  Canada  British Columbia  West Vancouver     V6V 8Z3   

       Gender             Education   Salary Marital Status Loyalty Card  \
2040     Male  High School or Below  81391.0        Married         Nova   
6123   Female              Bachelor  52435.0        Married         Nova   
10844  Female               College      NaN        Married         Star   
5291   Female                Master  81333.0         Single         Star   
11080    Male               College      NaN         Single         Star   

           CLV Enrollment Type  Enrollment Year  Enrollment Month  \
2040   5046.89        Standard             2014                11   
6123   7771.58        Standard             2015                 7   
10844  3883.40        Standard             2013                 8   
5291   6005.85        Standard             2012                 8   
11080  2030.78        Standard             2018                 5   

       Cancellation Year  Cancellation Month  
2040                 NaN                 NaN  
6123                 NaN                 NaN  
10844                NaN                 NaN  
5291                 NaN                 NaN  
11080                NaN                 NaN

# Fase 1: Exploración y Limpieza ✨

# Dimensiones df

In [95]:
#Analizamos las dimensiones de los dos dataframes

print(f'Num. filas df_vuelos: {df_vuelos.shape[0]}')
print(f'Num. columnas df_vuelos: {df_vuelos.shape[1]}')
print(f"\n---------------\n")
print(f'Num. filas df_clientes: {df_clientes.shape[0]}')
print(f'Num. columnas df_clientes: {df_clientes.shape[1]}')

Num. filas df_vuelos: 405624
Num. columnas df_vuelos: 10

---------------

Num. filas df_clientes: 16737
Num. columnas df_clientes: 16


# Nombres columnas

In [96]:
# Nombres de las columnas 
print(f'Nombres columnas df_vuelos: {df_vuelos.columns}')
print(f"\n---------------\n")
print(f'Nombres columnas df_clientes: {df_clientes.columns}')

Nombres columnas df_vuelos: Index(['Loyalty Number', 'Year', 'Month', 'Flights Booked',
       'Flights with Companions', 'Total Flights', 'Distance',
       'Points Accumulated', 'Points Redeemed', 'Dollar Cost Points Redeemed'],
      dtype='object')

---------------

Nombres columnas df_clientes: Index(['Loyalty Number', 'Country', 'Province', 'City', 'Postal Code',
       'Gender', 'Education', 'Salary', 'Marital Status', 'Loyalty Card',
       'CLV', 'Enrollment Type', 'Enrollment Year', 'Enrollment Month',
       'Cancellation Year', 'Cancellation Month'],
      dtype='object')


# Informacion df

In [97]:
# Informacion general de los df
print(f'Informacion df_vuelos: {df_vuelos.info()}')
print(f"\n---------------\n")
print(f'Informacion df_clientes: {df_clientes.info()}')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB
Informacion df_vuelos: None

---------------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data col

No tenemos variables categoricas en df_vuelos ✨

# % Nulos

In [98]:
nulos_vuelos = df_vuelos.isnull().sum()
nulos_clientes = df_clientes.isnull().sum()

# % Nulos
print(f'% Nulos df_vuelos: {(nulos_vuelos/df_vuelos.shape[0]*100).round(2)}')
print(f"\n---------------\n")
print(f'% Nulos df_clientes: {(nulos_clientes/df_clientes.shape[0]*100).round(2)}')

% Nulos df_vuelos: Loyalty Number                 0.0
Year                           0.0
Month                          0.0
Flights Booked                 0.0
Flights with Companions        0.0
Total Flights                  0.0
Distance                       0.0
Points Accumulated             0.0
Points Redeemed                0.0
Dollar Cost Points Redeemed    0.0
dtype: float64

---------------

% Nulos df_clientes: Loyalty Number         0.00
Country                0.00
Province               0.00
City                   0.00
Postal Code            0.00
Gender                 0.00
Education              0.00
Salary                25.32
Marital Status         0.00
Loyalty Card           0.00
CLV                    0.00
Enrollment Type        0.00
Enrollment Year        0.00
Enrollment Month       0.00
Cancellation Year     87.65
Cancellation Month    87.65
dtype: float64


- No tenemos valores nulos en df_vuelos ✨
- SI tenemos valores nulos en df_clientes 🚩 -> PDTE
    - Salary                25.32
    - Cancellation Year     87.65
    - Cancellation Month    87.65

# Duplicados 

In [99]:
print(f'Los duplicados que tenemos en df_vuelos: {df_vuelos.duplicated().sum()}')
print(f"\n---------------\n")
print(f'Los duplicados que tenemos en df_clientes: {df_clientes.duplicated().sum()}')

Los duplicados que tenemos en df_vuelos: 1864

---------------

Los duplicados que tenemos en df_clientes: 0


- No tenemos duplicados en df_clientes ✨
- SI tenemos duplicados en df_vuelos 🚩

# Limpieza duplicados df_vuelos

In [100]:
# Se que los duplicados estan en df_vuelos, verifico si solo estan duplicados una vez -> con keep=False cuenta todas las filas (originales y duplicadas)
df_vuelos.duplicated(keep=False).sum()

3712

In [101]:
# comprobamos duplicados en df_vuelos
df_vuelos[df_vuelos.duplicated(keep=False)].sort_values('Loyalty Number').head(15)

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
1006          101902  2017     12               0                        0   
1007          101902  2017     12               0                        0   
996           101902  2017      7               0                        0   
997           101902  2017      7               0                        0   
984           101902  2017      1               0                        0   
985           101902  2017      1               0                        0   
986           101902  2017      2               0                        0   
987           101902  2017      2               0                        0   
988           101902  2017      3               0                        0   
989           101902  2017      3               0                        0   
2770          106001  2018      6               0                        0   
2771          106001  2018      6               0                        0   
2995          106509  2017     10               0                        0   
3001          106509  2018      1               0                        0   
3000          106509  2018      1               0                        0   

      Total Flights  Distance  Points Accumulated  Points Redeemed  \
1006              0         0                 0.0                0   
1007              0         0                 0.0                0   
996               0         0                 0.0                0   
997               0         0                 0.0                0   
984               0         0                 0.0                0   
985               0         0                 0.0                0   
986               0         0                 0.0                0   
987               0         0                 0.0                0   
988               0         0                 0.0                0   
989               0         0                 0.0                0   
2770              0         0                 0.0                0   
2771              0         0                 0.0                0   
2995              0         0                 0.0                0   
3001              0         0                 0.0                0   
3000              0         0                 0.0                0   

      Dollar Cost Points Redeemed  
1006                            0  
1007                            0  
996                             0  
997                             0  
984                             0  
985                             0  
986                             0  
987                             0  
988                             0  
989                             0  
2770                            0  
2771                            0  
2995                            0  
3001                            0  
3000                            0

In [102]:
# Eliminaremos duplicados unicamente cuando todas las columnas sean idénticas, ya que para un mismo Loyalty Number nos muestra diferentes valores como el año en el cual se registraron las actividades de vuelo y el mes del año en el cual ocurrieron dichas actividades.
df_vuelos = df_vuelos.drop_duplicates()


In [103]:
#Analizamos las dimensiones de df_vuelos una vez eliminados los duplicados de filas exactas

print(f'Num. filas df_vuelos: {df_vuelos.shape[0]}')
print(f'Num. columnas df_vuelos: {df_vuelos.shape[1]}')
print(f"\n---------------\n")
print(f'Informacion df_vuelos: {df_vuelos.info()}')
print(f"\n---------------\n")
print(f'Los duplicados que tenemos en df_vuelos: {df_vuelos.duplicated().sum()}')

Num. filas df_vuelos: 403760
Num. columnas df_vuelos: 10

---------------

<class 'pandas.core.frame.DataFrame'>
Index: 403760 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               403760 non-null  int64  
 1   Year                         403760 non-null  int64  
 2   Month                        403760 non-null  int64  
 3   Flights Booked               403760 non-null  int64  
 4   Flights with Companions      403760 non-null  int64  
 5   Total Flights                403760 non-null  int64  
 6   Distance                     403760 non-null  int64  
 7   Points Accumulated           403760 non-null  float64
 8   Points Redeemed              403760 non-null  int64  
 9   Dollar Cost Points Redeemed  403760 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 33.9 MB
Informacion df_vuelos: None

---------------

Los duplicados

Se han eliminado un total de 1864 filas que contenian valores identicos, manteniendo las originales 

# Valores negativos?
- Lo comprobamos para df_vuelos teniendo en cuenta que todas sus columnas contienen valores numericos
- Lo comprobamos para las columnas con valores numericos de df_clientes

In [104]:
# df_vuelos
filas_negativas_df_vuelos = df_vuelos[(df_vuelos < 0).any(axis=1)]
display(filas_negativas_df_vuelos)


Empty DataFrame
Columns: [Loyalty Number, Year, Month, Flights Booked, Flights with Companions, Total Flights, Distance, Points Accumulated, Points Redeemed, Dollar Cost Points Redeemed]
Index: []

No tenemos valores negativos en df_vuelos ✨

In [105]:
# df_clientes
# Obtenemos lista con las columnas numericas
columnas_numericas_clientes = df_clientes.select_dtypes(include='number').columns.tolist()
print(columnas_numericas_clientes)
print(f"\n---------------\n")
# Compruebo si hay valores negativos en estas columnas
negativos_por_columna = (df_clientes[columnas_numericas_clientes] < 0).any()
print(negativos_por_columna)
print(f"\n---------------\n")
# La columna Salary tiene valores negativos, compruebo cuales son:
df_filas_negativas_salary = df_clientes[df_clientes['Salary'] < 0]
display(df_filas_negativas_salary.sort_values(['Salary']))


['Loyalty Number', 'Salary', 'CLV', 'Enrollment Year', 'Enrollment Month', 'Cancellation Year', 'Cancellation Month']

---------------

Loyalty Number        False
Salary                 True
CLV                   False
Enrollment Year       False
Enrollment Month      False
Cancellation Year     False
Cancellation Month    False
dtype: bool

---------------



Loyalty Number Country          Province            City Postal Code  \
952            152016  Canada           Ontario         Toronto     P1J 8T7   
15010          906428  Canada             Yukon      Whitehorse     Y2K 6R0   
8227           542976  Canada            Quebec        Montreal     H2Y 4R4   
2094           212128  Canada           Ontario         Toronto     P2T 6G3   
2570           239955  Canada            Quebec            Hull     J8Y 3Z5   
2445           232755  Canada  British Columbia       Vancouver     V1E 4R6   
6235           436087  Canada            Quebec        Montreal     H2T 9K8   
11830          734647  Canada      Saskatchewan          Regina     S1J 3C5   
7894           525245  Canada  British Columbia        Victoria     V10 6T5   
7277           491242  Canada  British Columbia    Dawson Creek     U5I 4F1   
4575           347013  Canada            Quebec     Quebec City     G1B 3L5   
12893          790475  Canada           Ontario         Trenton     K8V 4B2   
1728           194065  Canada           Ontario         Sudbury     M5V 1G5   
4888           364596  Canada            Quebec       Tremblant     H5Y 2S9   
11737          729561  Canada            Quebec     Quebec City     G1B 3L5   
9350           603070  Canada  British Columbia  West Vancouver     V6V 8Z3   
6120           430398  Canada      Newfoundland      St. John's     A1C 6H9   
16007          959977  Canada  British Columbia       Vancouver     V5R 1W3   
288            115505  Canada      Newfoundland      St. John's     A1C 6H9   
4925           366599  Canada           Ontario         Toronto     M1R 4K3   

       Gender             Education   Salary Marital Status Loyalty Card  \
952    Female              Bachelor -58486.0        Married       Aurora   
15010    Male              Bachelor -57297.0        Married         Star   
8227     Male  High School or Below -49830.0       Divorced         Star   
2094     Male              Bachelor -49001.0        Married         Nova   
2570   Female              Bachelor -47310.0        Married         Nova   
2445   Female              Bachelor -46683.0         Single         Nova   
6235     Male              Bachelor -46470.0        Married         Star   
11830    Male              Bachelor -46303.0        Married         Nova   
7894     Male              Bachelor -45962.0        Married         Star   
7277     Male              Bachelor -43234.0        Married         Star   
4575     Male              Bachelor -39503.0        Married         Nova   
12893  Female              Bachelor -34079.0        Married         Nova   
1728   Female              Bachelor -31911.0        Married         Nova   
4888   Female              Bachelor -26322.0         Single       Aurora   
11737  Female              Bachelor -19332.0       Divorced         Star   
9350   Female              Bachelor -19325.0         Single         Star   
6120     Male              Bachelor -17534.0        Married         Nova   
16007  Female              Bachelor -12497.0        Married       Aurora   
288      Male              Bachelor -10605.0        Married         Nova   
4925   Female              Bachelor  -9081.0        Married         Star   

            CLV Enrollment Type  Enrollment Year  Enrollment Month  \
952     5067.21  2018 Promotion             2018                 2   
15010  10018.66  2018 Promotion             2018                 4   
8227   24127.50  2018 Promotion             2018                 3   
2094    3130.68  2018 Promotion             2018                 2   
2570    6366.23  2018 Promotion             2018                 3   
2445    4787.81  2018 Promotion             2018                 3   
6235    4786.89  2018 Promotion             2018                 4   
11830  11280.73  2018 Promotion             2018                 4   
7894    2402.33  2018 Promotion             2018                 3   
7277    7597.91  2018 Promotion             2018  

In [106]:
print(f'Num filas en df_clientes que tienen valores negativos en Salary: {df_filas_negativas_salary.shape[0]}')


Num filas en df_clientes que tienen valores negativos en Salary: 20


Tenemos 20 valores negativos en la columna Salary de df_clientes 🚩 -> PDTE

# Limpieza valores negativos df_clientes

In [107]:
print(f'El % de valores negativos en df_salary es: {round((df_filas_negativas_salary.shape[0]/df_clientes.shape[0])*100,2)}')

El % de valores negativos en df_salary es: 0.12


El % de valores negativos es MUY BAJO -> 0.12 % 

# Principales estadisticos df_vuelos

In [108]:
# principales estadisticos de df_vuelos
display(df_vuelos.describe().T)
# NOTA: no tenemos variables categoricas en df_vuelos

count           mean            std       min  \
Loyalty Number               403760.0  549875.383713  258961.514684  100018.0   
Year                         403760.0    2017.500352       0.500000    2017.0   
Month                        403760.0       6.501335       3.451982       1.0   
Flights Booked               403760.0       4.134050       5.230064       0.0   
Flights with Companions      403760.0       1.036569       2.080472       0.0   
Total Flights                403760.0       5.170619       6.526858       0.0   
Distance                     403760.0    1214.460979    1434.098521       0.0   
Points Accumulated           403760.0     124.263761     146.696179       0.0   
Points Redeemed              403760.0      30.838587     125.758002       0.0   
Dollar Cost Points Redeemed  403760.0       2.495973      10.172033       0.0   

                                  25%       50%       75%       max  
Loyalty Number               326699.0  550598.0  772152.0  999986.0  
Year                           2017.0    2018.0    2018.0    2018.0  
Month                             4.0       7.0      10.0      12.0  
Flights Booked                    0.0       1.0       8.0      21.0  
Flights with Companions           0.0       0.0       1.0      11.0  
Total Flights                     0.0       1.0      10.0      32.0  
Distance                          0.0     525.0    2342.0    6293.0  
Points Accumulated                0.0      53.0     240.0     676.5  
Points Redeemed                   0.0       0.0       0.0     876.0  
Dollar Cost Points Redeemed       0.0       0.0       0.0      71.0

# Principales estadisticos df_clientes

In [109]:
# principales estadisticos de df_vuelos
display(df_clientes.describe().T.round(2))
print(f"\n---------------\n")
display(df_clientes.describe(include = "object").T)

count       mean        std        min        25%  \
Loyalty Number      16737.0  549735.88  258912.13  100018.00  326603.00   
Salary              12499.0   79245.61   35008.30  -58486.00   59246.50   
CLV                 16737.0    7988.90    6860.98    1898.01    3980.84   
Enrollment Year     16737.0    2015.25       1.98    2012.00    2014.00   
Enrollment Month    16737.0       6.67       3.40       1.00       4.00   
Cancellation Year    2067.0    2016.50       1.38    2013.00    2016.00   
Cancellation Month   2067.0       6.96       3.46       1.00       4.00   

                          50%        75%        max  
Loyalty Number      550434.00  772019.00  999986.00  
Salary               73455.00   88517.50  407228.00  
CLV                   5780.18    8940.58   83325.38  
Enrollment Year       2015.00    2017.00    2018.00  
Enrollment Month         7.00      10.00      12.00  
Cancellation Year     2017.00    2018.00    2018.00  
Cancellation Month       7.00      10.00      12.00


---------------



count unique       top   freq
Country          16737      1    Canada  16737
Province         16737     11   Ontario   5404
City             16737     29   Toronto   3351
Postal Code      16737     55   V6E 3D9    911
Gender           16737      2    Female   8410
Education        16737      5  Bachelor  10475
Marital Status   16737      3   Married   9735
Loyalty Card     16737      3      Star   7637
Enrollment Type  16737      2  Standard  15766

# Valores unicos de las variables categoricas de df_clientes

In [111]:
# creamos una lista con los nombres de las columnas categoricas 
columnas_cat_clientes = df_clientes.select_dtypes(include='object').columns.tolist()
print(columnas_cat_clientes)
# empezamos a iterar por cada una de las columnas para sacar sus valores únicos y sus frecuencias
for columna in columnas_cat_clientes:
    print(f" \n----------- ESTAMOS ANALIZANDO LA COLUMNA: '{columna.upper()}' -----------\n")
    print(f"Sus valores únicos son: {df_clientes[columna].unique()}\n")
    print(f"Las frecuencias de los valores únicos de las categorías son: {df_clientes[columna].value_counts()} ")

['Country', 'Province', 'City', 'Postal Code', 'Gender', 'Education', 'Marital Status', 'Loyalty Card', 'Enrollment Type']
 
----------- ESTAMOS ANALIZANDO LA COLUMNA: 'COUNTRY' -----------

Sus valores únicos son: ['Canada']

Las frecuencias de los valores únicos de las categorías son: Country
Canada    16737
Name: count, dtype: int64 
 
----------- ESTAMOS ANALIZANDO LA COLUMNA: 'PROVINCE' -----------

Sus valores únicos son: ['Alberta' 'Ontario' 'British Columbia' 'Saskatchewan' 'Quebec' 'Manitoba'
 'Newfoundland' 'New Brunswick' 'Nova Scotia' 'Prince Edward Island'
 'Yukon']

Las frecuencias de los valores únicos de las categorías son: Province
Ontario                 5404
British Columbia        4409
Quebec                  3300
Alberta                  969
Manitoba                 658
New Brunswick            636
Nova Scotia              518
Saskatchewan             409
Newfoundland             258
Yukon                    110
Prince Edward Island      66
Name: count, dtype: int6

No se observa ningun valor atipico en las variables categoricas de df_clientes ✨